### PR TITLE
Support RFC6570 URI Templates

### PR DESCRIPTION
**Pull request by alexdeem**
Remove `Path`, `Query`, `QueryMap`, `QueryName` parameter annotations; and replace with `Variable` which represents a variable for URI Template processing

### CODE REVIEW

**Comment by steineron** [comment]
@alexdeem looks very nice. Thank you.my two questions would be:
1. why forking and not branching? 
2. I'm wondering if there's any special reason to remove the ability for http queries.  as much as I loved adding the new capabilities - I'd also like to preserve any existing ones, just in case we'll need them, e.g when using some 3rd party service as we occasionally do.

File: retrofit/src/main/java/retrofit2/ParameterHandler.java [hunk 1]
```diff
@@ -80,100 +88,18 @@
...
-  }
-
-  static final class Query<T> extends ParameterHandler<T> {
+  static final class Variable<T> extends ParameterHandler<T> {
     private final String name;
     private final Converter<T, String> valueConverter;
-    private final boolean encoded;
 
-    Query(String name, Converter<T, String> valueConverter, boolean encoded) {
+    Variable(String name, Converter<T, String> valueConverter) {
```
Comment (steineron) [inline retrofit/src/main/java/retrofit2/ParameterHandler.java]: wouldn't the word "Expression" be a bit more aligned with the RFC here?

**Comment by alexdeem** [comment]
1. I can't make a branch in squares repo; we need to fork. We can look towards contributing this back, but it'd have to be for a future major version.
2. We haven't removed any functionality here; it's just achieved differently. The old school way is to use a template only to define the path of the endpoint; and we use different parameters to define the query. RFC6570 allows for templating the entirety of the URL including the query string.

**Comment by steineron** [comment]
Thanks @alexdeem  - I can appreciate taking the url expansion task with the like of RFC6570. My point was that we limit ourselves to use *only* this way of url-expansion  and should we require the old (and not so gold) way of paths+queries - we won't have it...
Perhaps a reasonable price to pay... I simply opt to leave capabilities when possible.

**Comment by alexdeem** [comment]
Again, we're not losing any capabilities; we're just changing the syntax for achieving it.
I chose this path explicitly to avoid having two different methods of achieving the same thing, removing the need for the client to explicitly choose between them.

**Pull request closed without merging**